### PR TITLE
Add summary of comments from Issue #73

### PR DESCRIPTION
**Pull request by tstellar**

### CODE REVIEW

File: rfc_code_review/comment_summary.md
Comment (joker-eph) [inline rfc_code_review/comment_summary.md:46]:
```suggestion
* E-mails don't include patch content; 
```

We rely on this for having the mailing list archives as a system of record. This was even a condition to accept Phabricator in the first place.

Comment (joker-eph) [reply rfc_code_review/comment_summary.md:46]:
```suggestion
* Not a single workflow: many options to do the same thing (for example: rebase+force-push or not?)
```

Not sure how to phrase it, but this is something that makes the overall list of missing feature a bit hard. Because for example we may say that GitHub does not support "Incremental review by looking at only the delta", but someone may say that you can -- if you never ever force-push. However in this case you can't rebase and need to merge main into the PR branch? 

So there is a list of things that we possibly won't have because we'll want to choose a specific workflow and our choice will make some things possible but others harder.

**Review by joker-eph — COMMENTED** [review]
Thanks, that's a good start. I left some comments inline.

File: rfc_code_review/comment_summary.md
Comment (ChristianKuehnel) [reply rfc_code_review/comment_summary.md:38]: If we have PRs and pre-merge testing, i would like to have an auto-submit option:
* The user sets a flag for auto-merge (e.g. via a comment or label)
* The PR is auto-merged, once some agreed conditions are met (e.g. pre-merge checks passing, at least n human approvals).

This would reduce the friction of pre-merge testing, especially for small patches.

File: rfc_code_review/comment_summary.md
Comment (ChristianKuehnel) [inline rfc_code_review/comment_summary.md:54]:
```suggestion
* Investigate alternative interfaces (https://pullrequest.com, reviewable.io)
* Start discussion around pre-merge testing and what it would take to make it mandatory
```

File: rfc_code_review/comment_summary.md
Comment (joker-eph) [reply rfc_code_review/comment_summary.md:38]: It's not clear in what you're saying that this helps new contributors who don't have commit access and so can't trigger the merge (neither through a button, and hopefully not with a comment!).
Also there a bunch of things up in the air in your answer: for example what will the commit message looks like if a PR is auto-merged?

> This would reduce the friction of pre-merge testing, especially for small patches.

I agree, I'm very supportive of a "fire-and-forget" submit queue being available. But that's not the claim here: there not mention of **testing** in this paragraph.

File: rfc_code_review/comment_summary.md
Comment (tstellar) [reply rfc_code_review/comment_summary.md:10]: There wasn't agreements on all these points in the issues, so I tried to give a general summary of what was discussed.  My goal was to emphasize the overall sentiment that Phabricator comments were more intuitive that GitHub.

File: rfc_code_review/comment_summary.md
Comment (tstellar) [reply rfc_code_review/comment_summary.md:16]: The second category is more for what happens to people who aren't subscribed to a pull requests.

File: rfc_code_review/comment_summary.md
Comment (tstellar) [reply rfc_code_review/comment_summary.md:19]: As someone with very little experience with pull requests, this was my take away from the overall discussion.   There was a lot of back and forth and different ideas, and for me it wasn't clear which one was right.  I also am trying to consider that the audience of this summary may also be people who are unfamiliar with pull requests, so I don't know if there is a lot of value in listing the various solutions for this.

File: rfc_code_review/comment_summary.md
Comment (tstellar) [reply rfc_code_review/comment_summary.md:26]: What's a better way to phrases this ?  s/Can/We could if we wanted to/

File: rfc_code_review/comment_summary.md
Comment (tstellar) [reply rfc_code_review/comment_summary.md:38]: >How is it "more intuitive for new contributors"? Since they can't merge the pull-request themselves anyway and will need someone to do it for them, it should be exactly identical?

This was mentioned in the comments on the issue.

File: rfc_code_review/comment_summary.md
Comment (tstellar) [reply rfc_code_review/comment_summary.md:46]: Are these suggestions for the Missing Features section  ?

File: rfc_code_review/comment_summary.md
Comment (tstellar) [reply rfc_code_review/comment_summary.md:54]: I think we should be doing this last point anyway even if we aren't moving to GitHub PRs.  I don't know if I want to link it to this.

**Comment by tstellar** [comment]
This seems like it has stalled a little bit, how can we move this forward?

File: rfc_code_review/comment_summary.md
Comment (joker-eph) [reply rfc_code_review/comment_summary.md:16]: Sorry, I still don't quite get it, can you elaborate?

File: rfc_code_review/comment_summary.md
Comment (joker-eph) [reply rfc_code_review/comment_summary.md:19]: Instead of "difficult to do", it seems more accurate to present it as "does not have an appropriate equivalent in GitHub".

File: rfc_code_review/comment_summary.md
Comment (joker-eph) [reply rfc_code_review/comment_summary.md:38]: That does not seem enough of a reason to me to just phrase it this way.

File: rfc_code_review/comment_summary.md
Comment (joker-eph) [reply rfc_code_review/comment_summary.md:26]: This can't be in a "GitHub positive Features" IMO: because this isn't part of the A/B comparison of the review tooling **within the bounds of our process**.

You may want to add another specific section like `## Others` at the end to add something like that indeed "If we wanted to change our workflow in the future, Pull-Requests would allow us to enforce that code contributions are going through mandatory automated checks" or something like that.

File: rfc_code_review/comment_summary.md
Comment (joker-eph) [reply rfc_code_review/comment_summary.md:10]: I still believe that this paragraph does not reflect appropriately the thread, but I can make a direct edit if you prefer?

Comment (tstellar) [reply rfc_code_review/comment_summary.md:10]: > I still believe that this paragraph does not reflect appropriately the thread, but I can make a direct edit if you prefer?

Yeah, if you want make direct edits or suggestions with that exact text you think would be better that would help a lot.

**Comment by joker-eph** [comment]
(somehow didn't send these, I just noticed this was "pending")

**Comment by tstellar** [comment]
@joker-eph Would it be easier if you just created a new version of the summary based off of my first draft?  My original idea was that we would have multiple people review the comments and then try to merge the summaries together.  So if you did a second version that would be closer to how I had envisioned this working from the beginning.  It's been over 6 weeks since I've reviewed the comments, and every time I try to make it an edit I feel like I have to re-read all the comments again.

**Comment by joker-eph** [comment]
It would have been better if you had clarified this early, can you just take all the edits right now and then we can merge this?

File: rfc_code_review/comment_summary.md
Comment (joker-eph) [reply rfc_code_review/comment_summary.md:16]: Ping on this?

File: rfc_code_review/comment_summary.md
Comment (joker-eph) [inline rfc_code_review/comment_summary.md:51]:
```suggestion
  * You can merge pull requests directly from the web UI, which avoid leaving the review tool
     to push to main, avoid having to manipulate command line tools (`arc` or `patch`, `git`).
     Phabricator does not have an equivalent.
```

File: rfc_code_review/comment_summary.md
Comment (joker-eph) [inline rfc_code_review/comment_summary.md:6]:
```suggestion
   * On a subjective level, some community members mentioned they are less
```

**Comment by tstellar** [comment]
> It would have been better if you had clarified this early, can you just take all the edits right now and then we can merge this?

This was in the mail I sent to the list: https://groups.google.com/u/1/a/llvm.org/g/iwg/c/LVf3Di-xOlM
I didn't realize you were going to read all the comments when you started reviewing this pull request, but I feel better now that someone else other than me has looked at it.

File: rfc_code_review/comment_summary.md
Comment (joker-eph) [inline rfc_code_review/comment_summary.md:34]:
```suggestion
* GitHub isn't open-source and does not offer us any way to implement missing features of customize it for the needs of the project.

## GitHub Positive Features
```

File: rfc_code_review/comment_summary.md
Comment (joker-eph) [inline rfc_code_review/comment_summary.md:60]:
```suggestion

```

File: rfc_code_review/comment_summary.md
Comment (joker-eph) [inline rfc_code_review/comment_summary.md:33]: * E-mails don't include patch content.

Comment (joker-eph) [inline rfc_code_review/comment_summary.md:33]:
```suggestion
* GitHub isn't open-source and does not offer us any way to implement missing features of customize it for the needs of the project.
* E-mails don't include patch content; 
* ```

File: rfc_code_review/comment_summary.md
Comment (joker-eph) [inline rfc_code_review/comment_summary.md:62]:
```suggestion

### GitHub Pull Request Recommended Action Items
```

File: rfc_code_review/comment_summary.md
Comment (joker-eph) [inline rfc_code_review/comment_summary.md:35]:
```suggestion
```

File: rfc_code_review/comment_summary.md
Comment (tstellar) [inline rfc_code_review/comment_summary.md:29]:
```suggestion
```

File: rfc_code_review/comment_summary.md
Comment (joker-eph) [inline rfc_code_review/comment_summary.md:61]:
```suggestion

### Workflow Opportunities with GitHub

* Can enforce developer policy / restrict pushes.
   * With GitHub pull requests we can enforce parts of the developer policy with automated checks.
* One point that was made in the comments is that it is hard to have reliable pre-commit testing with Phabricator,
     because Phabricator deals only with patches, so if the patch doesn't apply on the current main branch or some
     other known commit then you can't run CI on the patch.  With GitHub pull requests everything is a branch, so
     you can always run CI against them even if the patch is a little bit out-of-date with main.
```

File: rfc_code_review/comment_summary.md
Comment (joker-eph) [inline rfc_code_review/comment_summary.md:38]:
```suggestion
```

File: rfc_code_review/comment_summary.md
Comment (joker-eph) [inline rfc_code_review/comment_summary.md:44]:
```suggestion
```

**Review by joker-eph — APPROVED** [review]
LG right now, but there are two discussions opened (one from me and and one from Christian), we'll still need to follow up on this.

File: rfc_code_review/comment_summary.md
Comment (joker-eph) [inline rfc_code_review/comment_summary.md:5]:
```suggestion
* UI Deficiencies:
```

File: rfc_code_review/comment_summary.md
Comment (joker-eph) [inline rfc_code_review/comment_summary.md:6]:
```suggestion
  * On a subjective level, some community members mentioned they are less
```

File: rfc_code_review/comment_summary.md
Comment (joker-eph) [inline rfc_code_review/comment_summary.md:10]:
```suggestion
  * GitHub does not allow to comment on arbitrary line outside of the
```

File: rfc_code_review/comment_summary.md
Comment (joker-eph) [inline rfc_code_review/comment_summary.md:20]:
```suggestion
      immediate context of the changed lines.
  * Comments become hidden and discussion marked as "obsolete"
      automatically, even when the comment isn't actually addressed or
      the discussion isn't marked "resolved" by the author. In Phabricator
      hiding is an explicit action from each individual.
      This leads to a recurring complain of an impression of "disappearing
      comments": they disappear from the diff view, but remain in the
      history of the "Conversation" tab, disconnected from the patch.
  * In case of force-push (either amend or rebase):
    * the UI does not provide a direct access to view the changes since a
```

File: rfc_code_review/comment_summary.md
Comment (joker-eph) [inline rfc_code_review/comment_summary.md:31]:
```suggestion
* E-mails don't include patch content.
```

File: rfc_code_review/comment_summary.md
Comment (joker-eph) [inline rfc_code_review/comment_summary.md:56]:
```suggestion
    * With GitHub pull requests we can enforce parts of the developer policy with automated checks.
```

File: rfc_code_review/comment_summary.md
Comment (joker-eph) [inline rfc_code_review/comment_summary.md:22]:
```suggestion
    * Comments can't be viewed in their original context, you only see the
```

Comment (joker-eph) [inline rfc_code_review/comment_summary.md]:
```suggestion
    * the UI does not provide a direct access to view the changes since a
        previous state.
    * Comments can't be viewed in their original context, you only see the
```

**Comment by tstellar** [comment]
> LG right now, but there are two discussions opened (one from me and and one from Christian), we'll still need to follow up on this.

Which of your discussions is still opened?  Also, should I squash all commits and force push to the branch?

**Comment by joker-eph** [comment]
> > LG right now, but there are two discussions opened (one from me and and one from Christian), we'll still need to follow up on this.
> 
> Which of your discussions is still opened?

The fact that you're asking is a testimony of the "confusing" part of GitHub UI :)
There should be only one discussion that isn't marked "resolved", least that's what I see in the "conversation" tab (and it does not show up on the diff tab anymore, which is why people are complaining about "disappearing comments").
The first discussion thread I see in the conversion tabs ends up with a message from me "Ping on this?".

> Also, should I squash all commits and force push to the branch?

You may if you'd like. Note you don't need to: if you're unfamiliar, the "merge pull request" button has a little arrow nest to it, which allows to manage this:

Screen Shot 2021-12-23 at 9 42 30 PM

File: rfc_code_review/comment_summary.md
Comment (tstellar) [reply rfc_code_review/comment_summary.md]: I deleted the "Configurable Notifications" bullet point and kept the one that mentioned "Herald Rules".

**Comment by tstellar** [comment]
> > > LG right now, but there are two discussions opened (one from me and and one from Christian), we'll still need to follow up on this.
> > 
> > 
> > Which of your discussions is still opened?
> 
> The fact that you're asking is a testimony of the "confusing" part of GitHub UI :) There should be only one discussion that isn't marked "resolved", least that's what I see in the "conversation" tab (and it does not show up on the diff tab anymore, which is why people are complaining about "disappearing comments"). The first discussion thread I see in the conversion tabs ends up with a message from me "Ping on this?".
> 
I found it now, thanks.

> > Also, should I squash all commits and force push to the branch?
> 
> You may if you'd like. Note you don't need to: if you're unfamiliar, the "merge pull request" button has a little arrow nest to it, which allows to manage this:
> 
> Screen Shot 2021-12-23 at 9 42 30 PM

Does that let you edit the commit message before the changes get pushed?

**Comment by joker-eph** [comment]
> Does that let you edit the commit message before the changes get pushed?

Yes. That actually was discussed in the thread: we can't really review the final commit description because it depends what the person clicking the "merge button" will do with the box. By default it appends all the individual commit message (or an extract of each of them).

For example if I click "Squash and Merge" in the current state, here is the box:

Screen Shot 2021-12-23 at 11 06 24 PM

**Comment by tstellar** [comment]
@ChristianKuehnel Any more comments on this?

**Comment by ChristianKuehnel** [comment]
> @ChristianKuehnel Any more comments on this?

@tstellar 
Sorry I was on vacation and am buried in other work at the moment. Please proceed as you see fit.

**Comment by ChristianKuehnel** [comment]
@tstellar ping on merging the change (and probably notifying the community).

**Pull request merged**
